consensus/istanbul/backend/announce.md

```diff
@@ -0,0 +1,140 @@
+
+### p2p connection management
+
+Wether a validator opens a connection or not against another validator is not in the scope of this spec, but only how and when should the protocol messages should be sent, replied and regossipped. For example, it IS part of the spec what messages should be sent to a newly connected peer or validator peer.
+
+## Protocol Messages (RLP)
+
+// add types with the RLP nomenclature
+
+### queryEnodeMsg (0x12)
+
+`[[dest_address: B_20, encrypted_enode_url: B], version: P, timestamp: P]`
+
+- `dest_address`: the destination validator address this message is querying to.
+- `encrypted_enode_url`: the origin's `eNodeURL` encrypted with the destination public key.
+- `version`: the current announce version for the origin's `eNodeURL`.
+- `timestamp`: a message generation timestamp, used to bypass hash caches for messages.
+
+Max amount of encrypted node urls is 2 * size(set of [NearlyElectedValidator]).
+
+### enodeCertificateMsg (0x17)
+
+`[enode_url: B, version: P]`
+
+- `enode_url`: the origin's plaintext `eNodeURL`.
+- `version`: the current announce version for the origin's `eNodeURL`.
+
+### versionCertificatesMsg (0x16)
+
+`[[version: P, signature: B]]`
+
+- `version`: the current highest known announce version for a validator (deduced from the signature), according to the peer that gossipped this message.
+- `signature`: the signature for the version payload string (`'versionCertificate'|version`) by the emmitter of the certificate. Note that the address and public key can be deduced from this `signature`.
+
+## Nearly Elected Validator (NEV)
+
+A validator is a Nearly Elected Validator (NEV) if and only if:
+
+- It is a current elected validator in this epoch, or
+- It is a validator in the result set from calling `ElectNValidatorSigners` with `additionalAboveMaxElectable` taking the value of 10.
+
+## Spec by node type
+
+Note that validators are full nodes too, therefore they should follow the same rules as full nodes. This is specially important to not give away the status of the node as a validator.
+
+### Full node (non validator) Spec
+
+#### Peer handshake
+
+During an inbound peer connection, the remote peer can send an [enodeCertificateMsg] to identify itself as a validator. This allows for preferential treatment for the p2p connection.
+
+#### Peer registered
+
+When a peer is registered, all highest known version certificates should be sent to the registered peer in a [versionCertificatesMsg].
+
+#### Handling [queryEnodeMsg]
+
+Messages received of this type should be only processed once.
+Should be regossipped as is, unless another message from the same validator origin has been regossipped in the past 5 minutes.
+
+#### Handling [enodeCertificateMsg]
+
+[enodeCertificateMsg] should be ignored by non validators.
```

Comment:
I don't think non validators ever receive an enode certificate message, because they are only sent directly from validator to validator.

Reply:
They should never receive one, it should never happen, but there could be border cases where it happen, as in, a validator being restarted into a full node, or whatever. I'm thinking we might as well make that part explicit here.